apps/src/blockly/addons/plusMinusBlocks/advancedProcedures.js

```diff
@@ -690,3 +610,196 @@ const procedureVars = function () {
+  ) {
+    return;
+  }
+
+  const argName =
+    sourceArgBlock.outputConnection?.targetConnection?.getParentInput()?.name;
+  const workspace = sourceArgBlock.workspace;
+  const argData = defBlock.argData_;
+  const argDatum = defBlock.argData_.find(element => element.argId === argName);
+  const currId = argDatum.model.getId();
+
+  // Replace all whitespace with normal spaces, then trim.
+  newValue = newValue.replace(/[\s\xa0]+/g, ' ').trim();
+  const caselessName = newValue.toLowerCase();
+
+  /**
+   * Returns true if the given argDatum is associated with this field, or has
+   * a different caseless name than the argDatum associated with this field.
+   * @param {{model: GoogleBlockly.VariableModel, argId:string}} argDatum The
+   *     argDatum we want to make sure does not conflict with the argDatum
+   *     associated with this field.
+   * @returns {boolean} True if the given datum does not conflict with the
+   *     datum associated with this field.
+   * @this {GoogleBlockly.FieldTextInput}
+   */
+  const hasDifName = argDatum => {
+    // The field name (aka id) is always equal to the arg id.
+    return (
+      argDatum.argId === argName ||
+      caselessName !== argDatum.model.name.toLowerCase()
+    );
+  };
+
+  if (!newValue || !argData.every(hasDifName)) {
+    GoogleBlockly.Procedures.mutateCallers(sourceArgBlock.getRootBlock());
+    return null;
+  }
+
+  // Create new vars instead of renaming the old ones, so users can't
+  // accidentally rename/coalesce vars.
+  let model = workspace.getVariable(newValue, '');
+  if (!model) {
+    model = workspace.createVariable(newValue, '');
+  } else if (model.name !== newValue) {
+    // Blockly is case-insensitive so we have to update the var instead of
+    // creating a new one.
+    workspace.renameVariableById(model.getId(), newValue);
+  }
+  if (model.getId() !== currId) {
+    argDatum.model = model;
+  }
+  GoogleBlockly.Procedures.mutateCallers(sourceArgBlock.getRootBlock());
+  return newValue;
+}
+
+GoogleBlockly.Extensions.register('argument_reporter_validator', function () {
+  this.getField('VAR').setValidator(argumentReporterValidator);
+  this.onchange = event => {
+    const workspace = GoogleBlockly.Workspace.getById(event.workspaceId);
+    const eventBlock = workspace.getBlockById(event.blockId);
+    if (eventBlock?.type === BLOCK_TYPES.argumentReporter) {
+      switch (event.type) {
+        case GoogleBlockly.Events.BLOCK_FIELD_INTERMEDIATE_CHANGE:
+          workspace
+            .getAllBlocks()
+            .filter(
+              block =>
+                block.getRootBlock() === eventBlock?.getRootBlock() &&
+                block.type === BLOCK_TYPES.argumentReporter &&
+                block.getFieldValue('VAR') === event.oldValue
+            )
+            .forEach(matchingBlock =>
+              matchingBlock.getField('VAR').setValue(event.newValue, false)
+            );
+        // eslint-disable-next-line no-fallthrough
```

Comment:
why do we need this?

Reply:
The fall-through is intentional in this case because we always want to update the warning text (for any of these `event.type` cases). If it's a field change, we want to update the other blocks and the warning text. If it's a block move or block change, we just care about the warning text. Will add a comment!